src/scripts/release.zig

```diff
@@ -865,6 +933,87 @@ fn publish_python(shell: *Shell, info: VersionInfo) !void {
     });
 }
 
+fn publish_rust(shell: *Shell, info: VersionInfo) !void {
+    var section = try shell.open_section("publish rust");
+    defer section.close();
+
+    assert(try shell.dir_exists("zig-out/dist/rust"));
+
+    const token = try shell.env_get("CRATES_IO_TOKEN");
+
+    try shell.pushd("./src/clients/rust");
+    defer shell.popd();
+
+    const cargo_version = shell.exec_stdout("cargo --version", .{}) catch {
+        return error.NoRust;
+    };
+    log.info("{s}", .{cargo_version});
+
+    // `cargo check` will put all pre-compiled assets are in place.
+    // It will call `zig build clients:rust` itself.
+    // Do this before `cargo package`.
+    try shell.exec("cargo check", .{});
+
+    try backup_create(shell.cwd, "Cargo.toml");
+    defer backup_restore(shell.cwd, "Cargo.toml");
+    // Changeng the version will modify the lockfile.
+    try backup_create(shell.cwd, "Cargo.lock");
+    defer backup_restore(shell.cwd, "Cargo.lock");
+
+    const cargo_toml = try shell.cwd.readFileAlloc(
+        shell.arena.allocator(),
+        "Cargo.toml",
+        1 * MiB,
+    );
+    const version_line = try shell.fmt(
+        "version = \"{s}\"",
+        .{info.tag},
+    );
+    const cargo_toml_updated = try std.mem.replaceOwned(
+        u8,
+        shell.arena.allocator(),
+        cargo_toml,
+        "version = \"0.1.0\"",
+        version_line,
+    );
+    assert(std.mem.indexOf(u8, cargo_toml_updated, version_line) != null);
+
+    try shell.cwd.writeFile(.{
+        .sub_path = "Cargo.toml",
+        .data = cargo_toml_updated,
+    });
+
+    // --allow-dirty is needed because the rust crate is packaging
+    // static libraries that are not committed to git, and because
+    // the lockfile is dirty from changing the crate version.
+    try shell.exec("cargo publish --token {token} --allow-dirty", .{
+        .token = token,
+    });
+
+    // The cargo we publish with (rust 1.63) does not wait for
+    // the package to be available like modern cargo.
+    // For reliability of later release verification, we would like to wait.
+    // Curl can't do this in a one-liner sadly.
+    const curl_command = "curl --http2 -I -L -s https://crates.io/api/v1/crates/tigerbeetle/{version}/download";
+    const sleeps_max = 10;
+    const sleep_seconds = 5;
+    _ = for (0..sleeps_max) |_| {
+        log.info("waiting {}s for rust crate to be available", .{sleep_seconds});
+        std.time.sleep(sleep_seconds * std.time.ns_per_s);
+        const stdout = try shell.exec_stdout(curl_command, .{
+            .version = info.tag,
+        });
+        if (std.mem.indexOf(u8, stdout, "\nHTTP/2 403\n") != null) {
+            // 403 forbidden seems to be what crates.io returns for unpublished crates.
+            continue;
+        }
+        log.info("rust crate available on crates.io", .{});
+        break;
+    } else {
+        log.err("rust crate not yet available on crates.io", .{});
```

Comment:
Hm, let's maybe `return error.CargoPublishFailed` here? I feel we should _either_ guarantee that a crate is published here (so we return an error) _or_ embrace eventual consistency and add a retry loop in `verify_release` instead. 

Maybe the latter is better? Retry `cargo add` a couple of times?

Reply:
Oh, important context is that _some_ registers are eventually consistent (.NET and Java ones), so we don't necessarily gain much by forcing crates.io to be consistent. If we could get it for free, I'd take it, but, given that we want a workaround, I'd maybe just embrace it, and handle it on the consumer side.